standard/structs.md

```diff
@@ -38,12 +38,38 @@ struct_modifier
     | 'protected'
     | 'internal'
     | 'private'
+    | 'readonly'
     ;
 ```
 
 It is a compile-time error for the same modifier to appear multiple times in a struct declaration.
 
-The modifiers of a struct declaration have the same meaning as those of a class declaration ([§15.2.2](classes.md#1522-class-modifiers)).
+Except for `readonly`, the modifiers of a struct declaration have the same meaning as those of a class declaration ([§15.2.2](classes.md#1522-class-modifiers)).
+
+The `readonly` modifier indicates that the *struct_declaration* declares a type whose instances are immutable.
+
+A readonly struct has the following constraints:
+
+- Each of its instance fields shall also be declared `readonly`.
+- None of its instance properties shall have a *set_accessor_declaration* ([§15.7.3](classes.md#1573-accessors)).
+- It shall not declare any field-like events ([§15.8.2](classes.md#1582-field-like events)).
+
+When an instance of a readonly struct is passed to a method, its `this` is treated like an `in` argument/parameter, which disallows write access to any instance fields (except by constructors).
+
+### §ref-modifier-new-clause Ref modifier
+
+The `ref` modifier indicates that the *struct_declaration* declares a type whose instances are allocated on the execution stack.
+
+It is a compile-time error if a ref struct type is used in any of the following contexts:
+- As the element type of an array.
+- As the declared type of a field of a class or a non-ref struct.
+- To implement an interface.
+- Being boxed to `System.ValueType` or `System.Object`.
+- As a type argument.
+- An async method.
+- An iterator.
```

Comment:
Does "a ref struct type is used" include this case:

```C#
using System;
using System.Threading.Tasks;

class Widget
{
    public static implicit operator Widget(Span<int> s) => null;
}

static class Class1
{
    static async Task M(Task t)
    {
        Widget w = stackalloc int[] { 42 };
        await t;
    }
}
```

.NET SDK 6.0.101 compiles it without errors. The Span\<int\> becomes a local variable in the MoveNext method of the state machine.

Reply:
Resolving this conversation in this PR, as this concern should be covered in #601